tests/unit/fem/test_calcvshape.cpp

```diff
@@ -0,0 +1,437 @@
+
+   // Get a uniform grid or integration points
+   RefinedGeometry* ref = GlobGeometryRefiner.Refine( fe->GetGeomType(), res);
+   const IntegrationRule& intRule = ref->RefPts;
+
+   int npoints = intRule.GetNPoints();
+   for (int i=0; i < npoints; ++i)
+   {
+      // Get the current integration point from intRule
+      IntegrationPoint pt = intRule.IntPoint(i);
+
+      // Get several variants of this integration point
+      // some of which are inside the element and some are outside
+      Array<IntegrationPoint> ipArr;
+      GetRelatedIntegrationPoints( pt, dim, ipArr );
+
+      // For each such integration point check that the weights
+      // from CalcShape() sum to one
+      for (int j=0; j < ipArr.Size(); ++j)
+      {
+         IntegrationPoint& ip = ipArr[j];
+         fe->CalcVShape(ip, weights);
+
+         weights.MultTranspose(dofsx, v);
+         REQUIRE( v[0] == Approx(1.) );
+         if (dim > 1)
+         {
+            weights.MultTranspose(dofsy, v);
+            REQUIRE( v[1] == Approx(1.) );
+         }
+         if (dim > 2)
+         {
+            weights.MultTranspose(dofsz, v);
+            REQUIRE( v[2] == Approx(1.) );
+         }
+      }
+   }
+}
+
+/*
+TEST_CASE("CalcShape for several Lagrange FiniteElement instances",
+          "[Lagrange1DFiniteElement]"
+          "[BiLinear2DFiniteElement]"
+          "[BiQuad2DFiniteElement]"
+          "[LagrangeHexFiniteElement]")
+{
+   int maxOrder = 5;
+   int resolution = 10;
+
+   SECTION("Lagrange1DFiniteElement")
+   {
+      for (int order =1; order <= maxOrder; ++order)
+      {
+         std::cout << "Testing Lagrange1DFiniteElement::CalcShape() "
+                   << "for order " << order << std::endl;
+         Lagrange1DFiniteElement fe(order);
+         TestCalcShape(&fe, resolution);
+      }
+   }
+
+   SECTION("BiLinear2DFiniteElement")
+   {
+      std::cout << "Testing BiLinear2DFiniteElement::CalcShape()" << std::endl;
+      BiLinear2DFiniteElement fe;
+      TestCalcShape(&fe, resolution);
+   }
+
+   SECTION("BiQuad2DFiniteElement")
+   {
+      std::cout << "Testing BiQuad2DFiniteElement::CalcShape()" << std::endl;
+      BiQuad2DFiniteElement fe;
+      TestCalcShape(&fe, resolution);
+   }
+
+
+   SECTION("LagrangeHexFiniteElement")
+   {
+      std::cout << "Testing LagrangeHexFiniteElement::CalcShape() "
+                << "for order 2" << std::endl;
+
+      // Comments for LagrangeHexFiniteElement state
+      // that only degree 2 is functional for this class
+      LagrangeHexFiniteElement fe(2);
+      TestCalcShape(&fe, resolution);
+   }
+}
+*/
+TEST_CASE("CalcVShape for several ND FiniteElement instances",
+          "[ND_SegmentElement]"
+          "[ND_TriangleElement]"
+          "[ND_QuadrilateralElement]"
+          "[ND_TetrahedronElement]"
+          "[ND_WedgeElement]"
+          "[ND_HexahedronElement]")
+{
+   int maxOrder = 5;
+   int resolution = 10;
+
+   SECTION("ND_SegmentElement")
```

Comment:
Interesting you included the SegmentElement in this one but not the others?

Reply:
The `ND_SegmentElement` is an odd case. It can be evaluated as a 1D vector field aligned with the 1D domain and its `CalcVShape` function will populate a `DenseMatrix` containing only a single column. It has no `CalcCurlShape` function because this would be identically zero. I'm not sure how useful `ND_SegmentElement` is but it does have a `CaclVShape` function so we should test it.